### PR TITLE
New run glue script

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

This PR cleans up the `run_glue.py` script to use the Datasets library. Along the way it adds a few fixes in Trainer. The script supports all glue tasks as well as custom user tasks (passed along with a training and validation file in csv or json format). It has been tested on the following setups:
- single GPU
- multi-GPU with DataParallel
- multi-GPU with DistributedDataParallel
- TPU

The README has been updated to reflect the changes, there is just one breaking change from before which is that `data_dir` is not an accepted argument anymore (since Datasets will take care of downloading the data files).

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -812,7 +812,6 @@ def _maybe_log_save_evalute(self, tr_loss, model, trial, epoch):
         if self.control.should_evaluate:
             metrics = self.evaluate()
             self._report_to_hp_search(trial, epoch, metrics)
-            self.control = self.callback_handler.on_evaluate(self.args, self.state, self.control, metrics)
```
Comment (sgugger) [inline src/transformers/trainer.py:815]: Moving this to the end of `evaluate` othewise that even is not called when we call `trainer.evaluate()` independently.

File: examples/text-classification/run_glue.py [hunk 2]
```diff
@@ -15,30 +15,69 @@
...
+    AutoConfig,
+    AutoModelForSequenceClassification,
+    AutoTokenizer,
+    EvalPrediction,
     HfArgumentParser,
     Trainer,
     TrainingArguments,
-    glue_compute_metrics,
-    glue_output_modes,
-    glue_tasks_num_labels,
```
Comment (LysandreJik) [inline examples/text-classification/run_glue.py:36]: This is very nice

File: examples/text-classification/run_glue.py [hunk 3]
```diff
@@ -85,30 +132,21 @@ def main():
...
-        bool(training_args.local_rank != -1),
-        training_args.fp16,
+        f"Process rank: {training_args.local_rank}, device: {training_args.device}, n_gpu: {training_args.n_gpu}"
+        + f"distributed training: {bool(training_args.local_rank != -1)}, 16-bits training: {training_args.fp16}"
     )
-    logger.info("Training/evaluation parameters %s", training_args)
+    # Set the verbosity to info on main process only:
+    if is_main_process(training_args.local_rank):
+        logging.set_verbosity_info()
+    logger.info(f"Training/evaluation parameters {training_args}")
```
Comment (LysandreJik) [inline examples/text-classification/run_glue.py:143]: Have you confirmed this actually works? It seems to me that you're setting the default verbosity level of the *root* logger (so the loggers of `transformers` and every file contained in it), but the logger of the current file isn't a child of this logger (it's in `examples/`, not in `src/transformers/`)so it doesn't look like it'll be impacted by that change.

I would argue you would still need to change the current logger's default verbosity to `info` if you want to see the line `logger.info(f"Training/evaluation parameters {training_args}")` being printed.

**Review by LysandreJik — APPROVED** [review]
This is a fantastic change, it makes this example's code so much easier to read imo.

File: examples/text-classification/run_glue.py [hunk 4]
```diff
@@ -134,107 +173,98 @@ def main():
...
-        return compute_metrics_fn
+    datasets = load_dataset("glue", data_args.task_name)
+    sentence1_key, sentence2_key = task_to_keys[data_args.task_name]
+    encode_kwargs = {
+        "tokenizer": tokenizer,
+        "sentence1_key": sentence1_key,
+        "sentence2_key": sentence2_key,
+        "max_length": data_args.max_seq_length,
+    }
+    datasets = datasets.map(preprocess_function, batched=True, fn_kwargs=encode_kwargs)
```
Comment (thomwolf) [inline examples/text-classification/run_glue.py:184]: I like better to have `preprocess_function` as a closure written just here with the arguments instead of defining it above with kwargs.

It spares the reader a scroll up and down to see what's happening but I understand this is a matter of personal taste.

File: examples/text-classification/run_glue.py [hunk 5]
```diff
@@ -134,107 +173,98 @@ def main():
...
+
+    train_dataset = datasets["train"]
+    eval_dataset = datasets["validation_matched" if data_args.task_name == "mnli" else "validation"]
+    test_dataset = datasets["test_matched" if data_args.task_name == "mnli" else "test"]
+
+    # Get the metric function
+    metric = load_metric("glue", data_args.task_name)
+
+    def compute_metrics(p: EvalPrediction):
```
Comment (thomwolf) [inline examples/text-classification/run_glue.py:193]: Add a comment for the reader

File: examples/text-classification/run_glue.py [hunk 6]
```diff
@@ -134,107 +173,98 @@ def main():
...
-        def compute_metrics_fn(p: EvalPrediction):
-            preds = p.predictions[0] if isinstance(p.predictions, tuple) else p.predictions
-            if output_mode == "classification":
-                preds = np.argmax(preds, axis=1)
-            else:  # regression
-                preds = np.squeeze(preds)
-            return glue_compute_metrics(task_name, preds, p.label_ids)
-
-        return compute_metrics_fn
+    datasets = load_dataset("glue", data_args.task_name)
```
Comment (thomwolf) [inline examples/text-classification/run_glue.py:176]: Add a detailed multi-line comment explaining how the user can also easily load his own datasets as a JSON or CSV files (with mock examples) and linking to the relevant page of the `datasets` library.

File: examples/text-classification/run_glue.py [hunk 7]
```diff
@@ -134,107 +173,98 @@ def main():
...
+        "max_length": data_args.max_seq_length,
+    }
+    datasets = datasets.map(preprocess_function, batched=True, fn_kwargs=encode_kwargs)
+
+    train_dataset = datasets["train"]
+    eval_dataset = datasets["validation_matched" if data_args.task_name == "mnli" else "validation"]
+    test_dataset = datasets["test_matched" if data_args.task_name == "mnli" else "test"]
+
+    # Get the metric function
+    metric = load_metric("glue", data_args.task_name)
```
Comment (thomwolf) [inline examples/text-classification/run_glue.py:191]: Here we should also think about a user who would like to train on his own classification CSV dataset.

I think we should probably have a few "f1", "accuracy" metrics in `datasets` for such use cases. What do you think @LysandreJik @lhoestq @sgugger ?

File: examples/text-classification/run_glue.py (hunk 3 above)
Comment (sgugger) [reply examples/text-classification/run_glue.py:143]: It does work and I can see all the info being printed on my screen.

File: examples/text-classification/run_glue.py [hunk 8]
```diff
@@ -134,107 +173,98 @@ def main():
...
+        "tokenizer": tokenizer,
+        "sentence1_key": sentence1_key,
+        "sentence2_key": sentence2_key,
+        "max_length": data_args.max_seq_length,
+    }
+    datasets = datasets.map(preprocess_function, batched=True, fn_kwargs=encode_kwargs)
+
+    train_dataset = datasets["train"]
+    eval_dataset = datasets["validation_matched" if data_args.task_name == "mnli" else "validation"]
+    test_dataset = datasets["test_matched" if data_args.task_name == "mnli" else "test"]
```
Comment (thomwolf) [inline examples/text-classification/run_glue.py:188]: logger.info() a few dataset samples?

**Review by thomwolf — COMMENTED** [review]
Looks really cool!

A few user experience comments :)

File: examples/text-classification/run_glue.py (hunk 7 above)
Comment (sgugger) [reply examples/text-classification/run_glue.py:191]: I think all basic metrics provided by scikit-learn should be available in `datasets`, yes.

File: examples/text-classification/run_glue.py (hunk 4 above)
Comment (sgugger) [reply examples/text-classification/run_glue.py:184]: We lose the preprocessing caching for some reason when doing that.

File: examples/text-classification/run_glue.py (hunk 8 above)
Comment (sgugger) [reply examples/text-classification/run_glue.py:225]: Done

File: examples/text-classification/run_glue.py (hunk 6 above)
Comment (sgugger) [reply examples/text-classification/run_glue.py:203]: Done, I stayed basic on the examples since there is the link to the datasets documentation.

File: examples/text-classification/run_glue.py (hunk 5 above)
Comment (sgugger) [reply examples/text-classification/run_glue.py:236]: Done.

File: examples/text-classification/run_glue.py (hunk 3 above)
Comment (sgugger) [reply examples/text-classification/run_glue.py]: Ok, this should be working now.

File: src/transformers/trainer_utils.py [hunk 9]
```diff
@@ -151,3 +151,30 @@ class HPSearchBackend(ExplicitEnum):
     HPSearchBackend.OPTUNA: default_hp_space_optuna,
     HPSearchBackend.RAY: default_hp_space_ray,
 }
+
+
+def is_main_process(local_rank):
+    """
+    Whether or not the current process if the local process, based on `xm.get_ordinal()` (for TPUs) first, then on
```
Comment (julien-c) [inline src/transformers/trainer_utils.py:158]:
```suggestion
    Whether or not the current process is the local process, based on `xm.get_ordinal()` (for TPUs) first, then on
```

**Review by julien-c — APPROVED** [review]
nice job

**Comment by julien-c** [comment]
Should we start thinking about automating the creation of the metadata block for the model's model card?

here for instance we'd already have this info:
```
---
datasets:
- mrpc
metrics:
- f1
finetuned_from: bert-base-cased
---
```

**Comment by sgugger** [comment]
We could think of something like that and add a blank model card to be completed by the user in the final checkpoint. We could also include the results of the last evaluation if there is one.

**Review by LysandreJik — APPROVED** [review]
I think this is great. LGTM!

File: examples/text-classification/run_glue.py [hunk 10]
```diff
@@ -82,40 +153,69 @@ def main():
...
-    # Set seed
+    # Set seed before initializing model.
     set_seed(training_args.seed)
 
-    try:
-        num_labels = glue_tasks_num_labels[data_args.task_name]
-        output_mode = glue_output_modes[data_args.task_name]
-    except KeyError:
-        raise ValueError("Task not found: %s" % (data_args.task_name))
+    # Get the datasets
```
Comment (thomwolf) [inline examples/text-classification/run_glue.py:180]:
```suggestion
    # Get the datasets: you can either provide your own CSV/JSON training and evaluation files (see below)
    # or specify a GLUE benchmark task (the dataset will be downloaded automatically from the datasets Hub
    #
    # For CSV/JSON files, this script while use as label the column called 'label' and as pair of sentences the sentences in columns called 'sentence1' and 'sentence2' if such column exists or the first two columns if at least two columns are provided.
    # If the CSV/JSON contain only one non-label column, the script does single sentence classification on this single column.
    # You can easily tweak this behavior (see below)
```

File: examples/text-classification/run_glue.py [hunk 11]
```diff
@@ -82,40 +153,69 @@ def main():
...
+        # Loading a dataset from local csv files
+        datasets = load_dataset(
+            "csv", data_files={"train": data_args.train_file, "validation": data_args.validation_file}
+        )
+    else:
+        # Loading a dataset from local json files
+        datasets = load_dataset(
+            "json", data_files={"train": data_args.train_file, "validation": data_args.validation_file}
+        )
+    # See more about loading your dataset at https://huggingface.co/docs/datasets/loading_datasets.html.
```
Comment (thomwolf) [inline examples/text-classification/run_glue.py:197]:
```suggestion
    # See more about loading any type of standard or custom dataset at https://huggingface.co/docs/datasets/loading_datasets.html.
```

File: examples/text-classification/run_glue.py [hunk 12]
```diff
@@ -82,40 +153,69 @@ def main():
...
     )
-    logger.info("Training/evaluation parameters %s", training_args)
+    # Set the verbosity to info of the Transformers logger (on main process only):
+    if is_main_process(training_args.local_rank):
+        transformers.utils.logging.set_verbosity_info()
+    logger.info(f"Training/evaluation parameters {training_args}")
 
-    # Set seed
+    # Set seed before initializing model.
     set_seed(training_args.seed)
```
Comment (thomwolf) [inline examples/text-classification/run_glue.py:178]: Note that we also have a set_seed method in the `datasets` library.

File: examples/text-classification/run_glue.py [hunk 13]
```diff
@@ -82,40 +153,69 @@ def main():
...
+            num_labels = len(label_list)
+        else:
+            num_labels = 1
+    else:
+        # Trying to have good defaults here, don't hesitate to tweak to your needs.
+        labels = datasets["train"]["label"]
+        label_list = list(set(labels))
+        label_list.sort()  # list(set(xxx)) is not deterministic so we sort the labels to make it so.
+        num_labels = len(label_list)
+        is_regression = num_labels == 1
```
Comment (thomwolf) [inline examples/text-classification/run_glue.py:213]:
```suggestion
        is_regression = datasets['train'].features['label'].dtype in ['float32', 'float64']
        if is_regression:
            num_labels = 1
        else:
            # A useful fast method: https://huggingface.co/docs/datasets/package_reference/main_classes.html#datasets.Dataset.unique
            label_list = datasets["train"].unique("label").sort()  # Let's sort it for determinism
            num_labels = len(label_list)
```

File: examples/text-classification/run_glue.py [hunk 14]
```diff
@@ -133,108 +234,165 @@ def main():
...
+        # Again, we try to have some nice defaults but don't hesitate to tweak to your use case.
+        non_label_columns = datasets["train"].column_names
+        non_label_columns.remove("label")
+        if "sentence1" in non_label_columns and "sentence2" in non_label_columns:
+            sentence1_key, sentence2_key = "sentence1", "sentence2"
+        else:
+            if len(non_label_columns) >= 2:
+                sentence1_key, sentence2_key = non_label_columns[:2]
+            else:
+                sentence1_key, sentence2_key = non_label_columns[0], None
```
Comment (thomwolf) [inline examples/text-classification/run_glue.py:250]:
```suggestion
        non_label_column_names = [name for name in datasets["train"].column_names if name != "label"]
        if "sentence1" in non_label_column_names and "sentence2" in non_label_column_names:
            sentence1_key, sentence2_key = "sentence1", "sentence2"
        else:
            if len(non_label_column_names) >= 2:
                sentence1_key, sentence2_key = non_label_columns[:2]
            else:
                sentence1_key, sentence2_key = non_label_columns[0], None
```

File: examples/text-classification/run_glue.py [hunk 15]
```diff
@@ -133,108 +234,165 @@ def main():
...
+
+    # Padding strategy
+    if data_args.pad_to_max_length:
+        padding = "max_length"
+        max_length = data_args.max_seq_length
+    else:
+        # We will pad later at batch creation
```
Comment (thomwolf) [inline examples/text-classification/run_glue.py:257]:
```suggestion
        # We will pad later, dynamically at batch creation, to the max sequence length in each batch
```

**Review by thomwolf — APPROVED** [review]
Look great, added a few proposals to make it a bit simpler to read (imo)
